libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2017 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include "enums.h"
+#include "types.h"
+
+#include <string.h>
+#include <errno.h>
+
+#include <flatpak.h>
+#include <json-glib/json-glib.h>
+#include <ostree.h>
+
+#include "flatpak.h"
+#include "util.h"
+
+static FlatpakRemoteRefAction *
+flatpak_remote_ref_action_new (EosUpdaterUtilFlatpakRemoteRefActionType  type,
+                               FlatpakRemoteRef                         *ref,
+                               gint32                                    serial)
+{
+  FlatpakRemoteRefAction *action = g_slice_new0 (FlatpakRemoteRefAction);
+
+  action->ref_cnt = 1;
+  action->type = type;
+  action->ref = g_object_ref (ref);
+  action->serial = serial;
+
+  return action;
+}
+
+FlatpakRemoteRefAction *
+flatpak_remote_ref_action_ref (FlatpakRemoteRefAction *action)
+{
+  ++action->ref_cnt;
+  return action;
+}
+
+void
+flatpak_remote_ref_action_unref (FlatpakRemoteRefAction *action)
+{
+  if (--action->ref_cnt != 0)
```

Comment:
Add preconditions:
```
g_return_if_fail (action->ref_count > 0);
```
(this is the standard precondition for `unref()` functions)

Reply:
I'll add these given that they are the standard, but given that `action->ref_count` is `guint` I don't think they will serve much use?

Reply:
Helps in the double-unref case, when `action->ref_count == 0` at the start of the second call.